### ISSUE TITLE
关于源和目标RPN训练方面的问题

### ISSUE THREAD

**Issue by zpyovo** [chaoqichen/HTCN#2]
在训练过程中，我看到在对目标域的样本进行faster rcnn前向时候做了目标域的标记true，但是在faster-rcnn-HTCN.py代码中,  if target:的标记放在了PRN训练过程的后面
`def forward(self, im_data, im_info, gt_boxes, num_boxes,target=False,eta=1.0):
        batch_size = im_data.size(0)

        im_info = im_info.data
        gt_boxes = gt_boxes.data
        num_boxes = num_boxes.data

        # feed image data to base model to obtain base feature map
        base_feat1 = self.RCNN_base1(im_data)
        if self.lc:
            d_pixel, _ = self.netD_pixel(grad_reverse(base_feat1, lambd=eta))
            #print(d_pixel)
            # if not target:
            _, feat_pixel = self.netD_pixel(base_feat1.detach())
        else:
            d_pixel = self.netD_pixel(grad_reverse(base_feat1, lambd=eta))

        if self.la_attention:
            base_feat1 = local_attention(base_feat1, d_pixel.detach())

        base_feat2 = self.RCNN_base2(base_feat1)
        if self.gc:
            domain_mid, _ = self.netD_mid(grad_reverse(base_feat2, lambd=eta))
            # if not target:
            _, feat_mid = self.netD_mid(base_feat2.detach())
        else:
            domain_mid = self.netD_mid(grad_reverse(base_feat2, lambd=eta))

        if self.mid_attention:
            base_feat2 = middle_attention(base_feat2, domain_mid.detach())

        base_feat = self.RCNN_base3(base_feat2)
        if self.gc:
            domain_p, _ = self.netD(grad_reverse(base_feat, lambd=eta))
            # if target:
            #     return d_pixel,domain_p,domain_mid#, diff
            _,feat = self.netD(base_feat.detach())
        else:
            domain_p = self.netD(grad_reverse(base_feat, lambd=eta))

            # if target:
            #     return d_pixel,domain_p,domain_mid#,diff
        # feed base feature map tp RPN to obtain rois
        rois, rpn_loss_cls, rpn_loss_bbox = self.RCNN_rpn(base_feat, im_info, gt_boxes, num_boxes)

        # if it is training phrase, then use ground trubut bboxes for refining
        if self.training:
            roi_data = self.RCNN_proposal_target(rois, gt_boxes, num_boxes)
            rois, rois_label, rois_target, rois_inside_ws, rois_outside_ws = roi_data

            rois_label = Variable(rois_label.view(-1).long())
            rois_target = Variable(rois_target.view(-1, rois_target.size(2)))
            rois_inside_ws = Variable(rois_inside_ws.view(-1, rois_inside_ws.size(2)))
            rois_outside_ws = Variable(rois_outside_ws.view(-1, rois_outside_ws.size(2)))
        else:
            rois_label = None
            rois_target = None
            rois_inside_ws = None
            rois_outside_ws = None
            rpn_loss_cls = 0
            rpn_loss_bbox = 0

        rois = Variable(rois)
        # do roi pooling based on predicted rois

        if cfg.POOLING_MODE == 'align':
            pooled_feat = self.RCNN_roi_align(base_feat, rois.view(-1, 5))
        elif cfg.POOLING_MODE == 'pool':
            pooled_feat = self.RCNN_roi_pool(base_feat, rois.view(-1, 5))

        # feed pooled features to top model
        pooled_feat = self._head_to_tail(pooled_feat)
        #feat_pixel = torch.zeros(feat_pixel.size()).cuda()
        #
        feat_pixel = feat_pixel.view(1, -1).repeat(pooled_feat.size(0), 1)
        feat_mid = feat_mid.view(1, -1).repeat(pooled_feat.size(0), 1)
        feat = feat.view(1, -1).repeat(pooled_feat.size(0), 1)
        # concat
        feat = torch.cat((feat_mid, feat), 1)
        feat = torch.cat((feat_pixel, feat), 1)
        #
        feat_random = self.RandomLayer([pooled_feat, feat])

        d_ins = self.netD_da(grad_reverse(feat_random, lambd=eta))

        if target:
            return d_pixel, domain_p, domain_mid, d_ins`
而，RPN中也没有关于target数据的开关控制，那是不是意味着目标域的gt信息也进入了RPN的训练？
`# generating training labels and build the rpn loss
        if self.training:
            assert gt_boxes is not None

            rpn_data = self.RPN_anchor_target((rpn_cls_score.data, gt_boxes, im_info, num_boxes))

            # compute classification loss
            rpn_cls_score = rpn_cls_score_reshape.permute(0, 2, 3, 1).contiguous().view(batch_size, -1, 2)
            rpn_label = rpn_data[0].view(batch_size, -1)

            rpn_keep = Variable(rpn_label.view(-1).ne(-1).nonzero().view(-1))
            rpn_cls_score = torch.index_select(rpn_cls_score.view(-1,2), 0, rpn_keep)
            rpn_label = torch.index_select(rpn_label.view(-1), 0, rpn_keep.data)
            rpn_label = Variable(rpn_label.long())
            self.rpn_loss_cls = F.cross_entropy(rpn_cls_score, rpn_label)`

**Comment by chaoqichen** [comment]
@zpyovo 您好，感谢您的问题。这个不是这样的，输入目标域图像的时候，我们并没有加入rpn或faster rcnn的loss，只加入了对抗的loss，因此在输入目标域图像训练时，rpn模块并没有更新。

**Comment by chaoqichen** [comment]
@zpyovo 您好，感谢您的问题，这个不是您说的这样的。目标域信息没有引入到rpn中进行训练，因为输入目标域图像的时候，我们并没有加入rpn或faster-rcnn的loss，只加入了对抗的loss，因此在输入目标域图像训练时，rpn模块并没有更新。
